R-package/R/lgb.Dataset.R

```diff
@@ -45,25 +66,25 @@ Dataset <- R6::R6Class(
           stop("lgb.Dataset: If provided, predictor must be a ", sQuote("lgb.Predictor"))
       }
 
-      # Check for additional parameters
-      additional_params <- list(...)
-
-      # Check if attribute key is in the known attribute list
-      for (key in names(additional_params)) {
-
-        # Key existing
-        if (key %in% .INFO_KEYS()) {
-
-          # Store as info
-          info[[key]] <- additional_params[[key]]
-
-        } else {
-
-          # Store as param
-          params[[key]] <- additional_params[[key]]
-
-        }
```

Comment:
oh! I don't think so, but it's not obvious at all what I did, so let me share some links.

First, `Dataset` in the R package is not exported, so its constructor can be considered "private" in the package's API. That isn't true for any other public methods on this class, since they can be called by user code once a function like `lgb.Dataset()` returns a `Dataset`. So as long as all the exported methods / functions maintain backwards-compatibility, no user code will be broken.

Since this PR adds each of the elements of `INFO_KEYS` as keyword arguments in `Dataset$new()`, the only thing that could be left in `...` is information you want to include in params.

As of this PR, exported entrypoints that create a `Dataset` will handle merging `params` and `...` *before* calling `Dataset$new()`.

`Dataset$create_valid()`

https://github.com/microsoft/LightGBM/blob/c8212d812aacc697ea7ca3935c1adea73ac4edca/R-package/R/lgb.Dataset.R#L136-L156

`lgb.Dataset()`

https://github.com/microsoft/LightGBM/blob/c8212d812aacc697ea7ca3935c1adea73ac4edca/R-package/R/lgb.Dataset.R#L798-L826

=====

this made me realize that one place, `Dataset$slice()`, is missing the "combine params before calling `Dataset$new()` logic. I'll open a separate PR for that.

https://github.com/microsoft/LightGBM/blob/32445aba964f8e8f86a63fb2601f39a35863fd2d/R-package/R/lgb.Dataset.R#L501-L527

Reply:
> As of this PR, exported entrypoints that create a `Dataset` will handle merging `params` and `...` *before* calling `Dataset$new()`.

Ah, here is the key point! Thanks a lot for as always detailed explanation!